### ISSUE TITLE
Gate 0 RED: K=2 two-circle outer cascade rejects all 12 seeds at startup validation (recorded run, A100 box, HEAD 19ad612)

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1003]
First recorded full run of `tests/sae_manifold_joint_two_circle_recovery.rs` (release, Azure A100 box, 48 min): **FAILED before the solver ever started**.

```
outer cascade must complete: REML smoothing optimization failed to converge:
no candidate seeds passed outer startup validation (SAE manifold two-circle recovery):
  generated=12, screened=12, exact_validated=12, solver_started=0
  rejection breakdown: rejected_by_kkt=8, rejected_by_domain=0, rejected_by_objective=0, rejected_by_budget=0, rejected_other=4
```

Two failure classes across the 12 seeds:

1. **8 seeds** — `pre-fit identifiability audit: decoder atom 'circle_0' has rank-0 weighted design (n=600, M_k=3); all assignment weights vanish or the basis is degenerate`. Note `π_0 = 1` in `ibp_stick_breaking_prior`, and the fixture's residual seed logits give atom 0 positive logits on its own circle's rows, so the gate product σ(l/τ)·π_0 cannot vanish from the prior side — suspicion falls on the **basis/coordinate state at the seed ρ being evaluated** (e.g. extreme per-atom ARD precision seeds collapsing t → degenerate Φ rows), or on the audit's rank tolerance at small gates.
2. **4 seeds** — `reml_criterion: arrow-Schur: per-row H_tt^(0) Cholesky failed: row 0 H_tt is non-PD at base ridge 0e0; evidence mode preserves the genuine Cholesky of H_tt and does not condition non-PD blocks` — strict evidence-mode rejection (the #748 policy) at the seed state.

Structural observations:

- The L1 collapse guards (#976) reseed a collapsed atom only **after an accepted outer iteration**; the pre-fit audit fires **before** the first iteration, so a cold seed with a vacuous atom is fatal-by-audit instead of reseed-able. The guard and the audit disagree about whose job a dead-at-seed atom is.
- The seed cascade burned its entire budget (12/12) on two structural causes; `rejected_by_kkt=8` classifies the audit failures, which is itself misleading (no KKT was ever evaluated).
- Contrast: a real-data fit (Qwen2.5-0.5B L12 activations, n=2000, K=4, ibp_map) passes startup on the same build — real activations give every atom seed mass, the planted two-circle cold start does not. So the wall is specific to structured cold starts, exactly the #997 / #969 cure surface.

Fix directions to evaluate (in cure order):
1. Run `enforce_active_mass_guard`'s reseed-once path (or an entry-specific equivalent) **before** the pre-fit audit during seed-cost evaluation, so a vacuous-at-seed atom is reseeded rather than fatal — this is the guard's stated contract extended to iteration 0.
2. Audit the seed generator's ρ ranges for the SAE objective: per-atom ARD precision seeds that collapse all coordinates produce degenerate Φ for **every** atom topology; such seeds are structurally vacuous for this model class and should be screened, not exact-validated.
3. The 4 non-PD evidence rejections are correct policy individually, but the cascade should distinguish 'seed is bad' from 'model class cannot start' — uniform structural rejection already early-exits; it should also surface WHICH ρ axis drove the degeneracy.

Full log on the box: `~/gate0.log` (azuser@a100box). Reproduce: `cargo test --release --test sae_manifold_joint_two_circle_recovery`.

**Comment by SauersML** [comment]
Root-cause notes from read-only analysis of current `main`.

1. Seed source and exact fixture values

The 12 outer seeds are produced by `crate::seeding::generate_rho_candidates` as called from `run_outer_with_plan` in `src/solver/outer_strategy.rs`. The fixture builds:

- `SaeManifoldRho::new(0, 0, vec![zeros(0); K])`
- `init_rho.to_flat()` length 2
- `OuterProblem::new(2).with_initial_rho([0, 0])`

So the outer vector is only:

`[log_lambda_sparse, log_lambda_smooth]`

There are no ARD coordinates in this fixture. With default `SeedConfig` (`bounds=(-12,12)`, `max_seeds=12`, risk profile `GeneralizedLinear`) and no heuristic lambdas, `generate_rho_candidates(2, None, default)` emits, before truncation, the GL primary/neutral/oversmoothed seeds, GL symmetric baselines, coordinate probes, pair probes, and then global sweeps. The list is truncated to 12. The explicit initial rho `[0,0]` is already seed 2, so it does not add another seed.

Exact generated order for this fixture:

0: [1, 1]
1: [0, 0]
2: [12, 12]
3: [2, 2]
4: [4, 4]
5: [-2, -2]
6: [-1.25, 1]
7: [3.25, 1]
8: [1, -1.25]
9: [1, 3.25]
10: [3, -1]
11: [-1, 3]

2. Why `circle_0` gates can vanish at seed-cost evaluation

The raw fixture seed state should not have atom 0 vanish from the IBP prior side. `AssignmentMode::ibp_map(TAU=0.5, ALPHA=1.0, learnable_alpha=false)` gives `z_k = sigmoid(logit_k / tau) * pi_k`, and `pi_0 = 1`. Because `learnable_alpha=false`, the outer `log_lambda_sparse` coordinate does not change `alpha`; it only weights the IBP assignment prior in the objective. `log_lambda_smooth` changes decoder smoothing. This fixture has no ARD rho entries, so there are no extreme ARD seeds either.

Therefore the 12 rho seeds do not directly transform logits, coordinates, tau, alpha, or ARD before the audit. If atom 0 is rank-0 at the audit, the audited term has already been mutated by the per-seed inner entry path, not by `from_flat()` itself. Current `main` declares `SaeManifoldOuterObjective::requires_continuation_path_entry() == true`, so every seed enters through `ContinuationPath`; that path couples a heavy rho walk with a tau leg (`4.0 -> 0.5`) and inner solves before the final seed-cost/evidence eval. That continuation/inner solve can drive all logits for an atom far negative; for IBP, `stable_logistic(logit/tau)` then becomes tiny or exactly 0 for sufficiently negative logits. At the raw fixture seed, though, atom 0's prior mass is not the explanation.

3. Rank-0 audit tolerance

`finalize_decoder_identifiability_audit` accumulates `G_k = (diag(a_k) Phi_k)^T (diag(a_k) Phi_k)` and calls `identifiability_audit::rank_of_gram`. That computes singular values as `sqrt(eigenvalues(G_k))`, then applies the same rank convention as RRQR:

`tol = 100 * f64::EPSILON * max(n_total, p, 1) * max(leading_singular, 1)`

For this audit, `n_total=600`, `p=M_k=3`, so the floor is about `1.33e-11` when `leading_singular < 1`. This means the audit can report rank 0 with small-but-nonzero gates if every singular value of `diag(a_k) Phi_k` is below about `1e-11`. It is not an exact `a_k == 0` test. For a circle basis with normal-size `Phi`, the gate scale only has to fall to roughly `1e-12` order before all singular values can be below tolerance. Exact underflow is not required.

4. Does `8c258dc53` cure the 8 rank-0 rejections?

Yes for the reported failure class. `8c258dc53` adds `self.enforce_active_mass_guard(0)?` in `run_joint_fit_arrow_schur` after basis/frame setup and before `finalize_decoder_identifiability_audit`. That guard checks each atom's max assignment mass and, if it is below `SAE_ATOM_ACTIVE_MASS_FLOOR`, reseeds that atom's logits once. For IBP-MAP the reseed sets the collapsed atom's logits to `0.0`, so atom 0 gets `sigmoid(0/tau) * pi_0 = 0.5` and the weighted design is no longer rank-0 unless the basis itself is truly degenerate. That is precisely the missing iteration-0 use of the #976 guard. It cures the 8 audit rejections; it does not change the evidence/log-det policy.

5. What cures the 4 non-PD evidence rejections

Those are a separate evidence-path problem. The message is from strict ridge-0 evidence factorization of per-row `H_tt` at a non-PD/off-optimum state. The correct cure is not the active-mass guard. Current `main` has the relevant evidence cure in `SaeManifoldTerm::reml_criterion_with_cache` / `reml_criterion_streaming_exact`: both route through `converge_inner_for_undamped_logdet` before taking the undamped (`ridge_t=0`, Direct) Arrow-Schur factor. Blame points that routing/commentary to `208f49dc7` (`fix(SAE #841,#842,#843,#844,#847,#850,#853,#857): manifold correctness + streaming logdet convergence`). The code explicitly documents the failure mode: an initial or partially stopped iterate can have an indefinite/rank-deficient per-row block, while the converged optimum should have PD `H_tt` for the exact Laplace log-det.

So the split is:

- `8c258dc53`: cures the 8 rank-0 weighted-design startup audit failures by running the active-mass reseed guard before the audit.
- `208f49dc7`-class evidence convergence path: cures the 4 non-PD evidence rejections by refining the inner `(t, beta)` state to KKT/step convergence before the strict ridge-0 log-det factor. If a seed still reaches a genuinely non-PD `H_tt` after that convergence driver, rejection remains correct policy.

No cargo was run for this analysis.

**Comment by SauersML** [comment]
Root-cause notes from read-only inspection of `main` for #1003.

## 1. The 12 outer rho seed candidates

For this fixture, `run_production_fit` builds:

```rust
SaeManifoldRho::new(SPARSITY.ln(), SMOOTHNESS.ln(), vec![Array1::<f64>::zeros(0); K])
```

with `SPARSITY = 1`, `SMOOTHNESS = 1`, `K = 2`, so the flat outer vector is exactly two coordinates:

```text
[log_lambda_sparse, log_lambda_smooth] = [0, 0]
```

There are no ARD coordinates in this fixture: each per-atom `log_ard` block has length 0.

`OuterProblem::new(n_params).with_initial_rho(init_rho_flat)` calls `run_outer_with_plan`, which first calls:

```rust
crate::seeding::generate_rho_candidates(cap.n_params, config.heuristic_lambdas.as_deref(), &config.seed_config)
```

using default `SeedConfig`: bounds `[-12, 12]`, `max_seeds = 12`, `risk_profile = GeneralizedLinear`. Since the explicit initial `[0, 0]` is already generated, it is not inserted again.

For `num_penalties = 2`, no heuristic lambdas, and GLM risk shift `+1`, the generated/truncated list is:

```text
0: [ 1.00,  1.00]   primary = zero anchor + GLM risk shift
1: [ 0.00,  0.00]   neutral baseline / explicit initial rho
2: [12.00, 12.00]   GLM over-smoothed upper-bound seed
3: [ 2.00,  2.00]   broad symmetric baseline
4: [ 4.00,  4.00]   broad symmetric baseline
5: [-2.00, -2.00]   broad symmetric baseline
6: [-1.25,  1.00]   coord-0 probe around primary, step 2.25
7: [ 3.25,  1.00]   coord-0 probe around primary, step 2.25
8: [ 1.00, -1.25]   coord-1 probe around primary, step 2.25
9: [ 1.00,  3.25]   coord-1 probe around primary, step 2.25
10:[ 3.00, -1.00]   sparse pair probe, step 2.0
11:[-1.00,  3.00]   sparse pair probe, step 2.0
```

The later global-shift candidates (`[3,3]`, `[5,5]`, etc.) are generated after these but are dropped by `max_seeds = 12`.

## 2. Why atom 0 can audit as rank-0 despite pi_0 = 1

The IBP-MAP gate map itself does not explain a literal atom-0 zero from the cold fixture:

```rust
z_ik = stable_logistic(logit_ik / temperature) * pi_k
```

and `ibp_stick_breaking_prior(K=2, alpha=1)` gives `pi_0 = 1`, `pi_1 = 1/2`. The fixture uses `AssignmentMode::ibp_map(TAU=0.5, ALPHA=1.0, learnable_alpha=false)`, so the outer `rho` seeds do not change `tau`, `alpha`, or `pi_k`. They also do not include ARD coordinates here.

At entry to `SaeManifoldOuterObjective::evaluate`, `baseline_rho.from_flat(rho_flat)` only maps the two seed coordinates to `log_lambda_sparse` and `log_lambda_smooth`. Then `reml_criterion` calls `run_joint_fit_arrow_schur`. The pre-fit decoder audit is at the top of `run_joint_fit_arrow_schur`, before the Newton loop advances temperature, assembles `H_tt`, or applies any coordinate/logit step. So an individual cold seed rho does not directly transform logits or coordinates before the audit.

The important mechanism is stateful seed evaluation: `SaeManifoldOuterObjective` owns a mutable `term`, and evaluating a seed runs an inner joint solve that can move assignment logits. A collapsed atom can therefore be present at the next entry/audit if the objective state was already driven there by seed-cost/proxy work or by a prior leg of the startup path. The recorded failure is consistent with that: the audit is seeing the current term state, not re-computing gates from the original residual-seed logits.

## 3. The rank-0 wording can cover tiny nonzero gates

`finalize_decoder_identifiability_audit` accumulates per-atom Grams for:

```text
D_k = diag(a_.k) * Phi_k
G_k = D_k^T D_k
```

then calls `rank_of_gram(G_k, n_total)`. `rank_of_gram` takes eigenvalues of `G_k`, square-roots them into singular values, and counts singular values above:

```text
tol = 100 * f64::EPSILON * max(n_total, M_k, 1) * max(leading_singular_value, 1)
```

For this fixture (`n_total = 600`, `M_k = 3`), if `leading_singular_value < 1`, the absolute floor is about `1.33e-11`. Thus a design with small-but-nonzero gates can still be counted as rank 0 if all singular values are below that floor. The error text says “all assignment weights vanish or the basis is degenerate”; numerically it means “the weighted design has no singular direction above the audit tolerance.”

## 4. Effect of commit 8c258dc53

Commit `8c258dc53` inserts:

```rust
self.enforce_active_mass_guard(0)?;
```

immediately before the pre-fit audit in `run_joint_fit_arrow_schur`.

That should cure the 8 rank-0 audit rejections whose cause is collapsed gate support. The guard checks each atom’s maximum assignment mass over rows against `SAE_ATOM_ACTIVE_MASS_FLOOR = 1e-3`; for IBP-MAP, a collapsed atom is reseeded by setting that atom’s logits to `0.0` on every row. With `tau=0.5`, that gives `sigmoid(0/tau)=0.5`, so atom 0 re-enters with gate `0.5*pi_0 = 0.5` and atom 1 would re-enter with `0.5*pi_1 = 0.25`. Genuinely degenerate basis rows would still fail the audit, which is correct.

I did not run cargo per the read-only/no-cargo instruction, so this is a code-path conclusion rather than a test result.

## 5. The 4 non-PD evidence rejections need a different cure

The non-PD class is not fixed by the active-mass guard. It occurs later in evidence/log-det mode, where `ArrowSolveOptions::direct().with_ill_conditioning_tolerated()` intentionally accepts ill-conditioned-but-PD blocks but refuses a genuinely non-PD `H_tt` at base ridge 0. The error text is explicit: evidence mode must not silently add ridge, because that would change the reported `log|H_tt|`.

So the cure is not “ridge the evidence factor.” The cure is to ensure the inner state is actually moved to a zero-ridge-PD optimum before the evidence factor is requested, or to treat that seed as bad and continue. The most direct code-level cure would be in `converge_inner_for_undamped_logdet`: when the zero-ridge evidence factor returns `PerRowFactorFailed` for non-PD `H_tt`, run another damped LM refinement through `run_joint_fit_arrow_schur` and retry the undamped factor, up to the existing refinement budget. Only if the block remains non-PD after refinement should the seed be rejected. That preserves evidence honesty while avoiding a fatal off-optimum logdet probe.

The newer continuation-entry path in `outer_strategy.rs` is also aligned with this cure direction: start SAE joint fits through a heavier-smoothing homotopy so the seed arrives at the target rho from a PD basin rather than asking the cold state for an immediate exact evidence factor.

**Comment by SauersML** [comment]
## Closed: gate-0 two-circle fit no longer rejects seeds; certified entry replaces the blind cascade

The "all 12 seeds rejected at startup validation" failure is resolved across four landed fixes, and the gate-0 fixture now reaches its assertions via the certified curvature-homotopy entry walk with **zero reseeds**.

### Root causes and fixes

1. **Entry-time active-mass guard** (`8c258dc53`). A cold seed can hand the fit an atom whose gates are vacuous on every row — ρ states the seeding heuristics never saw. The pre-fit identifiability audit then reported that atom as a fatal rank-0 weighted design and the seed died. Fix: run the `#976` active-mass reseed guard at iteration 0, **before** the audit, so a support collapse is answered with one observable re-seed instead of a rejection. (`enforce_active_mass_guard(0)` at fit entry.)

2. **Non-PD evidence factor retry** (`933b82641`). The undamped (ridge-0) evidence factorization could hit a non-PD pivot at a near-degenerate seed state and reject the whole seed. Fix: retry damped refinement before rejecting, so a transiently indefinite `H+Sλ` at the seed is recovered rather than fatal.

3. **State hygiene — verified already correct.** Codex flagged a suspected cross-seed leak of collapsed gate state through the objective's mutable term. Audited and confirmed **not present**: `SaeManifoldOuterObjective::reset()` does `self.term = self.baseline_term.clone()`, fully restoring the pristine baseline (gates, decoder, coords, and now the η dial + walk report) before each seed evaluation. Both the screening cascade (`rank_indices_with_budget_cascade` calls `obj.reset()` per `(stage, cap, idx)`) and the main seed loop (`obj.reset()` at the seed-body top) are state-hygienic. The new η-walk runs immediately after that reset, so cross-seed leakage cannot occur. No additional fix was needed; the warm-start *across outer iterations within a seed* stays intentional.

4. **Certified curvature-homotopy entry leg** (#1007). Rather than depend on the blind multi-seed multistart to stumble into the basin, the SAE objective now enters every seed through the predictor-corrector η-walk from the Eckart-Young anchor (η=0, global by construction) to the full curved basis (η=1). The walk holds the arrow-factor min-pivot positivity invariant with adaptive step-halving and records any bifurcation; on the gate-0 fixture it arrives on the certified branch with zero reseeds.

### Verification

`sae_two_circle_curvature_homotopy_entry_arrives_zero_reseed` (in `tests/sae_manifold_joint_two_circle_recovery.rs`) asserts the walk arrives at η=1 with **zero reseeds, no bifurcation, no inner collapse events**, and post-walk reconstruction R²≥0.9 — the gate-0 zero-reseed requirement. The pre-existing `sae_manifold_joint_two_circle_recovery_ibp_map` continues to assert end-to-end recovery (no-collapse active mass, plane σ_min, circular-coord recovery, R², determinism).

Closing as fixed.
